File: JavaScript/sliding_window/buy_and_sell_stock.js
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock

// Optimized Solution
// Sliding window with two pointers
// Time Complexity: O(n)
// Space Complexity: O(1)

/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  let l = 0;
  let r = 1;
  let maxP = 0;

  while (r < prices.length) {
    let profit = prices[r] - prices[l];
    if (profit >= 0) {
      maxP = maxP > profit ? maxP : profit;
    } else {
      l = r;
    }
    r++;
  }

  return maxP;
};

File: Python/sliding_window/buy_and_sell_stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        l, r = 0, 1
        maxP = 0

        while r < len(prices):
            profit = prices[r] - prices[l]
            if profit >= 0:
                maxP = max(maxP, profit)
            else:
                l = r
            r += 1

        return maxP
